Source/Core/VideoBackends/D3D12/D3DUtil.cpp

```diff
@@ -0,0 +1,1051 @@
+	HDC hDC = CreateCompatibleDC(nullptr);
+	HBITMAP hbmBitmap = CreateDIBSection(hDC, &bmi, DIB_RGB_COLORS, reinterpret_cast<void**>(&pBitmapBits), nullptr, 0);
+	SetMapMode(hDC, MM_TEXT);
+
+	// create a GDI font
+	HFONT hFont = CreateFont(24, 0, 0, 0, FW_NORMAL, FALSE,
+							FALSE, FALSE, DEFAULT_CHARSET, OUT_DEFAULT_PRECIS,
+							CLIP_DEFAULT_PRECIS, PROOF_QUALITY,
+							VARIABLE_PITCH, _T("Tahoma"));
+
+	if (nullptr == hFont)
+		return E_FAIL;
+
+	HGDIOBJ hOldbmBitmap = SelectObject(hDC, hbmBitmap);
+	HGDIOBJ hOldFont = SelectObject(hDC, hFont);
+
+	// Set text properties
+	SetTextColor(hDC, 0xFFFFFF);
+	SetBkColor  (hDC, 0);
+	SetTextAlign(hDC, TA_TOP);
+
+	TEXTMETRICW tm;
+	GetTextMetricsW(hDC, &tm);
+	m_line_height = tm.tmHeight;
+
+	// Loop through all printable characters and output them to the bitmap
+	// Meanwhile, keep track of the corresponding tex coords for each character.
+	int x = 0, y = 0;
+	char str[2] = "\0";
+	for (int c = 0; c < 127 - 32; c++)
+	{
+		str[0] = c + 32;
+		SIZE size;
+		GetTextExtentPoint32A(hDC, str, 1, &size);
+		if ((int)(x + size.cx + 1) > m_tex_width)
+		{
+			x  = 0;
+			y += m_line_height;
+		}
+
+		ExtTextOutA(hDC, x+1, y+0, ETO_OPAQUE | ETO_CLIPPED, nullptr, str, 1, nullptr);
+		m_tex_coords[c][0] = ((float)(x + 0))/m_tex_width;
+		m_tex_coords[c][1] = ((float)(y + 0))/m_tex_height;
+		m_tex_coords[c][2] = ((float)(x + 0 + size.cx))/m_tex_width;
+		m_tex_coords[c][3] = ((float)(y + 0 + size.cy))/m_tex_height;
+
+		x += size.cx + 3;  // 3 to work around annoying ij conflict (part of the j ends up with the i)
+	}
+
+	// Create a new texture for the font
+	// possible optimization: store the converted data in a buffer and fill the texture on creation.
+	// That way, we can use a static texture
+	std::unique_ptr<byte[]> texInitialData(new byte[m_tex_width * m_tex_height * 4]);
+
+	for (y = 0; y < m_tex_height; y++)
+	{
+		u32* pDst32_12 = reinterpret_cast<u32*>(static_cast<u8*>(texInitialData.get()) + y * m_tex_width * 4);
+		for (x = 0; x < m_tex_width; x++)
+		{
+			const u8 bAlpha = (pBitmapBits[m_tex_width * y + x] & 0xff);
+
+			*pDst32_12++ = (((bAlpha << 4) | bAlpha) << 24) | 0xFFFFFF;
+		}
+	}
+
+	CheckHR(
+		D3D::device12->CreateCommittedResource(
+		&CD3DX12_HEAP_PROPERTIES(D3D12_HEAP_TYPE_DEFAULT),
+		D3D12_HEAP_FLAG_NONE,
+		&CD3DX12_RESOURCE_DESC::Tex2D(DXGI_FORMAT_R8G8B8A8_UNORM, m_tex_width, m_tex_height, 1, 1),
+		D3D12_RESOURCE_STATE_COMMON,
+		nullptr,
+		IID_PPV_ARGS(&m_texture12)
+		)
+		);
+
+	D3D::SetDebugObjectName12(m_texture12, "texture of a CD3DFont object");
+
+	ID3D12Resource* temporaryFontTextureUploadBuffer;
+	CheckHR(
+		D3D::device12->CreateCommittedResource(
+		&CD3DX12_HEAP_PROPERTIES(D3D12_HEAP_TYPE_UPLOAD),
+		D3D12_HEAP_FLAG_NONE,
+		&CD3DX12_RESOURCE_DESC::Buffer(D3D12_TEXTURE_DATA_PLACEMENT_ALIGNMENT + m_tex_height * ((m_tex_width * 4 + D3D12_TEXTURE_DATA_PITCH_ALIGNMENT - 1) & ~(D3D12_TEXTURE_DATA_PITCH_ALIGNMENT - 1))),
```

Comment:
I'm not sure if this alignment has to be applied here, if the buffer you're creating is aligned to an equal or higher value, anyway. The only information I can find about this particular constants is for non-first sub-resource alignment.